### DISCUSSION TITLE
Specifying schema while exporting as parquet files

### DISCUSSION THREAD

**Discussion by ntapsrigowri** [duckdb/duckdb#4857] [Q&A]
I'm trying to generate a parquet file based on DuckDB data and upload it to Athena, BigQuery and Azure Synapse. The tables created in these 3rd party tools have specific schemas defined, is there an easy way to define and parquet schema for all the fields in the parquet file?

In the example below, i'd like to have uniqueId, file ,extension, ctime and owner as required fields. I would also like to have ctime as TIMESTAMP_MILLIS(DATETIME2 in Azure Synapse)
```

import duckdb from 'duckdb';

const changesTable = 'CREATE TABLE IF NOT EXISTS changes(file VARCHAR, extension VARCHAR, owner BIGINT);'

await runQuery(changesTable);

const insertChanges = "COPY changes FROM 'changes.csv' (AUTO_DETECT TRUE);"
await runQuery(insertChanges);

const parquetFile = COPY (SELECT ${Date.now()} as uniqueId, COALESCE(file,'') AS file, COALESCE(extension,'') AS extension, COALESCE(owner,0) AS owner FROM changes) TO 'changesschema.parquet' (FORMAT 'PARQUET');
await runQuery(parquetFile);

const parquetFileSchema = "select * from parquet_schema('changesschema.parquet');"
await runQuery(parquetFileSchema);

async function runQuery(query) {
const db = new duckdb.Database("changesDuck");
const response = await new Promise(resolve => {
db.all(query, (err, queryResponse) => {
if (err) {
console.log('Failed to run duck db query', query, err);
}
resolve(queryResponse);
});

});

console.log("Query RESPONS", query, response);
}
```

```
Response:
Query RESPONS CREATE TABLE IF NOT EXISTS changes(file VARCHAR, extension VARCHAR, owner BIGINT); []
Query RESPONS COPY changes FROM 'changes.csv' (AUTO_DETECT TRUE); [ { Count: 3 } ]
Query RESPONS COPY (SELECT COALESCE(file,''), COALESCE(extension,''), COALESCE(owner,0), CAST(TO_TIMESTAMP(ctime/1000) AS DATETIME) FROM changes) TO 'changesschema.parquet' (FORMAT 'PARQUET'); [ { Count: 12 } ]
Query RESPONS select * from parquet_schema('changesschema.parquet'); [
{
file_name: 'changesschema.parquet',
name: 'duckdb_schema',
type: 'BOOLEAN',
type_length: '0',
repetition_type: 'REQUIRED',
num_children: 3,
converted_type: 'UTF8',
scale: 0,
precision: 0,
field_id: 0,
logical_type: null
},
{
file_name: 'changesschema.parquet',
name: "COALESCE(file, '')",
type: 'BYTE_ARRAY',
type_length: '0',
repetition_type: 'OPTIONAL',
num_children: 0,
converted_type: 'UTF8',
scale: 0,
precision: 0,
field_id: 0,
logical_type: null
},
{
file_name: 'changesschema.parquet',
name: COALESCE("extension", ''),
type: 'BYTE_ARRAY',
type_length: '0',
repetition_type: 'OPTIONAL',
num_children: 0,
converted_type: 'UTF8',
scale: 0,
precision: 0,
field_id: 0,
logical_type: null
},
{
file_name: 'changesschema.parquet',
name: 'COALESCE("owner", 0)',
type: 'INT64',
type_length: '0',
repetition_type: 'OPTIONAL',
num_children: 0,
converted_type: 'INT_64',
scale: 0,
precision: 0,
field_id: 0,
logical_type: null
},
  {
    file_name: 'changesschema.parquet',
    name: 'ctime',
    type: 'INT64',
    type_length: '0',
    repetition_type: 'OPTIONAL',
    num_children: 0,
    converted_type: 'TIMESTAMP_MICROS',
    scale: 0,
    precision: 0,
    field_id: 0,
    logical_type: 'TimestampType(isAdjustedToUTC=0, unit=TimeUnit(MILLIS=<null>, MICROS=MicroSeconds(), NANOS=<null>))'
  }
]
```
GCP BigQuery Error:
Provided Schema does not match Table .changes. Field uniqueId has changed mode from REQUIRED to NULLABLE

Azure Synapse Error:
"Column 'ctime' of type 'DATETIME2' is not compatible with external data type 'Parquet physical type: INT64', please try with 'BIGINT'.

**Comment by hannes** [comment]
I'm still not entirely sure what you want to achieve. As a suggestion, pasting large amounts of outputs at us is not particularly helpful. Can you try to boil this down to a minimal reproducible example that clearly demonstrates where DuckDB is doing something wrong please? As it stands, it does not like something is problematic from our side.

**Reply by ntapsrigowri** [reply]
I pasted the code snippet to make it easier to run the test script :)
Let me make it precise. 

- I'm exporting data from DB as a parquet file. When I do that, i'm not able to specify a certain field as required(repetition_type: REQUIRED). 
- I would like to export data as TIMESTAMP_MILLIS in parquet file; i tried  `SELECT CAST(TO_TIMESTAMP(ctime/1000) AS DATETIME)` , but it returns type as TIMESTAMP_MICROS, how do we convert it to TIMESTAMP_MILLIS

I don't think DuckDB is doing anything wrong,I'm trying to understand how i could achieve the above scenarios

**Reply by hannes** [reply]
Can you elaborate on why you want the field to be `REQUIRED`?

**Reply by ntapsrigowri** [reply]
I'm trying to upload the parquet file to BigQuery, but get back this error ```Provided Schema does not match Table .changes. Field uniqueId has changed mode from REQUIRED to NULLABLE```

Similarly when i upload duckDB generated query on Azure Synapse, I see ```"Column 'ctime' of type 'DATETIME2' is not compatible with external data type 'Parquet physical type: INT64', please try with 'BIGINT'.```

**Reply by ntapsrigowri** [reply]
Any recommendation on how I could proceed? Is there a way we can specify a parquet file template in the export as parquet command?
Something, like, 
`COPY (SELECT* FROM changes) TO 'changesschema.parquet' (FORMAT 'PARQUET',SCHEMA CHANGES_SCHEMA); ` and we could define CHANGES_SCHEMA such that dataype, required/nullable, encoding can all be defined in the schema. 
For example: ParquetSchema  in https://github.com/ironSource/parquetjs

**Reply by ntapsrigowri** [reply]
Hi @hannes  , Just writing in to check if you have some suggestions or workarounds? Sorry about the follow up, I am pressed against time and I wanted to understand if the ask is something that we can accomplish with DuckDb

**Reply by Mytherin** [reply]
There is currently no support for defining a Parquet schema in the writer yourself. Perhaps export the data to Arrow and use their writer for now?

**Reply by ntapsrigowri** [reply]
Ok, thanks..I'll give it a try, is there a nodejs doc link that you could point me to , i could only see Python based doc to export data to arrow

**Reply by Mytherin** [reply]
Arrow export is not supported in the nodejs client yet, you would need to use the Python client for that.

**Reply by ntapsrigowri** [reply]
Thank you.. by the way, in the current state of things, we cannot mark a field as a `required` field as well is it?

**Reply by Mytherin** [reply]
DuckDB supports `NOT NULL` constraints in tables, but this information is not propagated into the Parquet writer yet.

**Reply by vachu** [reply]
> DuckDB supports `NOT NULL` constraints in tables, but this information is not propagated into the Parquet writer yet.

This propagation of NULL constraint ("NULLABLE" / "NOT NULL"), to the exported Parquet file, is atleast crucial for me.  Is there any initiative to implement this?
